argon2_out_of_process_support.go

```diff
@@ -0,0 +1,990 @@
+	}
+
+	// Start the remote process.
+	if err := cmd.Start(); err != nil {
+		return nil, fmt.Errorf("cannot start handler process: %w", err)
+	}
+
+	var actualRsp *Argon2OutOfProcessResponse // The response to return to the caller
+	exitWaitCh := make(chan struct{})         // A channel which signals successful exit of the child process when closed
+	tmb := new(tomb.Tomb)                     // To track all goroutines
+
+	// Spin up a routine to bootstrap the parent side and handle responses from
+	// the remote process.
+	tmb.Go(func() error {
+		// Spin up a routine for killing the child process if it doesn't
+		// die cleanly
+		tmb.Go(func() error {
+			<-tmb.Dying() // Wait here until the tomb enters a dying state
+
+			select {
+			case <-exitWaitCh:
+				// The command closed cleanly, so there's nothing to do.
+				return tomb.ErrDying
+			case <-time.NewTimer(5 * time.Second).C:
+				// We've waited 5 seconds - kill the child process instead. Go 1.20
+				// has a new feature (WaitDelay) which might make things a bit better
+				// here because I don't know how racey things are here - exec.Cmd is
+				// quite complicated.
+				if err := cmd.Process.Kill(); err != nil {
+					if err != os.ErrProcessDone {
+						return fmt.Errorf("failed to kill blocked remote process: %w", err)
+					}
+				}
+				return errors.New("killed blocked remote process")
+			}
+		})
+
+		// wdRspChan is sent watchdog responses on this goroutine, received from
+		// the remote process via stdout, and they are subsequently received by
+		// the watchdog monitor for processing.
+		wdRspChan := make(chan *Argon2OutOfProcessResponse)
+
+		// reqChan is sent the initial request from this goroutine and watchdog requests
+		// from the watchdog routine, which are received by a dedicated goroutine to
+		// serialize then and sends them to the remote process via its stdin.
+		reqChan := make(chan *Argon2OutOfProcessRequest)
+
+		// Spin up a routine for sending requests to the remote process via stdinPipe.
+		tmb.Go(func() error {
+			// Run a loop to send requests as long as the tomb is alive.
+			for tmb.Alive() {
+				select {
+				case req := <-reqChan:
+					// Send the request to the remote process via its stdin
+					enc := json.NewEncoder(stdinPipe)
+					if err := enc.Encode(req); err != nil {
+						return fmt.Errorf("cannot encode request: %w", err)
+					}
+				case <-tmb.Dying():
+					// The tomb is dying, so this loop will stop iterating.
+				}
+			}
+			return tomb.ErrDying
+		})
+
+		// Send the main request before starting the watchdog or running the response loop
+		select {
+		case reqChan <- req: // Unbuffered channel, but read end is always there unless the tomb is dying.
+		case <-tmb.Dying():
+			// The tomb has begun dying before we had a chance to send the initial request.
+			return tomb.ErrDying
+		}
+
+		// Spin up another routine to run the watchdog
+		tmb.Go(func() error {
+			err := k.watchdog(tmb, reqChan, wdRspChan)
```

Comment:
Watchdog adds quite a bit of complexity. I don't know the spec, but what are we trying to guard against here? if the watchdog handler needs to run a goroutine it doesn't prove anything more than that the child process is still running, but we would get that information from a process dying on its own. Maybe we could split the PR into a part without watchdog and then one adding it separately?

Reply:
It tests the loop in the remote process that encodes and sends responses back to the parent, so we have some confidence that we're going to receive a response back eventually (considering that this will be programmed to run for ~2s by default) - as long as we don't hit a bug in the crypto library that causes things to stall, of course - there's not much we can do for that. There might be in the future, eg, I'd like to make it interruptible (so you could pass around a `context.Context`), which would almost certainly mean forking the existing library. It wouldn't be difficult to work on, given that it's well specced. It's possible in this case that we could extend the API to add a progress channel to it, which would mean the watchdog could actually directly test that the crypto part is still executing.